docs/rfc/0696-p2p-node.adoc

```diff
@@ -0,0 +1,316 @@
+stats = {
+  "connections-total" ^ => uint,
+  "connected-peers" ^ => [* connect-info],
+  "membership-active" ^ => uint,
+  "membership-passive" ^ => uint,
+  * tstr => any
+}
+
+connect-info = [
+  peer_id,
+  [address],
+]
+
+address = address-ipv4 / address-ipv6
+address-ipv4 = [
+  ip4,
+  port,
+]
+address-ipv6 = [
+  ip6,
+  port,
+]
+
+ip4 = bstr .size 4
+ip6 = bstr .size 16
+
+; Canonical representation of a peer.
+peer-id: bstr
+
+; Network port.
+port: uint
+----
+
+All types representing requests and responses and their serialisation logic MUST
+be exposed as linkable libraries. It is RECOMMENDED to also expose the
+functionality to communicate with the node via IPC as a library.
+
+== Operations
+
+=== Supervision
+
+Process supervision SHOULD be deferred to established system level service
+managers i.e. `<<systemd>>` and `<<launchd>>` for Linux and macOS respectively.
+To support both long-running as well as ad-hoc usage the daemon implementation
+SHALL be equipped with the ability to detect and read the information from its
+environment necessary to determine if it has been activated via socket. When
+binding to a socket it SHALL use the file descriptors provided by the init
+process. If none are provided it SHALL assume long-running operation and SHALL
+bind to the well-known path on a UNIX domain socket in mode `SOCK_STREAM` under
+Linux:
+
+  $XDG_RUNTIME_DIR/radicle/<srv>-<peer-id>.sock
+
+macOS:
+
+  TODO(xla): Define well-known path for darwin systems.
+
+If sufficiently focused library implementations are present in the ecosystem to
+account for socket activation on the targeted platforms, this document
+RECOMMENDS using/contributing to those options.
+
+Both service managers offer support to fullfil the legacy `inetd` interface.
+Which is deemed insufficient for concenrs over security, lack of support for
+UNIX domain sockets and the design focusing on a process per connection.
+
+==== systemd
+
+Socket activation under systemd is passed on via: 
+
+* `LISTEN_PID` - MUST be equal to the PID of the daemon.
+* `LISTEN_FDS` - Number of received file descriptors, starting at 3.
+* `LISTEN_NAMES` - Contains colon-separated list of names corresponding to the
+  `FileDescriptorName` option in the service file.
+
+==== launchd
+
+* `LAUNCH_DAEMON_SOCKET_NAME` - Name of the socket according to the `.plist`
+  configuration file.
+
+The name passed to the process MUST be used to check-in with launchd as
+documented in `launch(3)` which in essence involves obtaining the FDs via
+`launch_activate_socket` expecting a name.
+
+=== Configuration
+
+Common service configuration files SHALL be provided alongside the source code
+of the node binary. To support the semi-dynamic nature of one process per
+profile, facilities are required to manage service definitions along the
+lifetime of a profile and when updated need to inform the process management
+system. Such facilities need to be considered wherever profiles are modified,
+i.e. CLI.
```

Comment:
@FintanH The author assumes this is significant for the work happening in #698 as it is tight to the profile management. Maybe you can chip in if this should be expanded on in a follow-up RFC which assumes presence of a cli and a daemon or if you see other ways of how this could work.

Reply:
It is of the opinion of the CLI author that there may be two routes we can go:
* Profile setting can only be switched if there is no peer-to-peer daemon running
* Or profile setting triggers a reload of the peer-to-peer daemon if it is currently running. Possibly through the PubSub component.

Perhaps there are other options, but these two are the ones that come to mind right now.

Reply:
I guess an alternative option is to have a way to spin up a separate peer-to-peer daemon for each profile. Maybe have the well known socket name of the peer to peer daemon include some indication of the profile ID.

Reply:
Mmmm that's a good point. `PeerId`s and `Profile`s have a 1:1 relationship, but then the whole idea of "active profile" becomes a bit murky.

Reply:
Can we make "active profile" an application concern? E.g have it be an environment variable for the CLI, a file in `<XDG_HOME>/radicle-upstream` for Upstream and so on.

Reply:
Very possible -- the changes here https://github.com/radicle-dev/radicle-link/pull/720 are all in terms of `RadHome` not relying on the active profile. There is a `PROFILE_ID` env var as well, which is currently only considered in `load`, but I've also outlined that [the profile should be a global parameter for subcommand](https://github.com/radicle-dev/radicle-link/pull/698/files#diff-a944b0f221c56c1f789fb8ba1ad84ba9c859dcf7a09c68430b9224769eb538f4R106).

Reply:
Some clarifications as I think we missed my original point in the ensued discussion. The assumption is that there is a 1:1 relation ship between profile and daemon. Truthfully 1:2 as there could be a node and a seed for any given monorepo. Which means that any system service manager must be configured in a way to run these. If at the same time or not is out of the control of those managers as the corresponding sockets can be activated at any point with no order guarantee - distributed systems in a box, ya know.

With all that in mind, when a profile is created a corresponding configuration for the service managers must be created (detailing how and where to run it) and fed to said managers can have them ready for operation. How the interplay is coming about between the templates of such service configuration, the rendering to an actual file and loading into the service manager is where I'd like to steer the conversation towards. And as I said above that can also happen in a follow-up and we assume for now that such configuration templates exist.

Reply:
> Truthfully 1:2 as there could be a node and a seed for any given monorepo.

That's not recommended :sweat_smile:  https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/0682-application-architecture.adoc#peer-to-peer-node
> we RECOMMEND to treat both as separate services, each with their own peer id and persistent state

> With all that in mind, when a profile is created a corresponding configuration (detailing how and where to run it) for the service managers must be created and fed to said managers so it can have them ready for operation.

Ah so for example, we might have `~/.config/radicle-link/42-69-3/protocol/membership.toml`, generated from some template? Is that what you're trying to discuss?

Reply:
> That's not recommended sweat_smile https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/0682-application-architecture.adoc#peer-to-peer-node

Thanks for the reminder!

> Ah so for example, we might have `~/.config/radicle-link/42-69-3/protocol/membership.toml`, generated from some template? Is that what you're trying to discuss?

Not really, more like `~/.config/systemd/user/linkd-<peer-id>.socket` and the corresponding invocations to `systemctl reload-daemon` or equivalents on other systems.

Reply:
Hmmm so what does this have to do with profiles? :thinking: Sorry if I'm slow with the uptake here :bow:

Reply:
The existence of a daemon is bound to the existence of a profile, as they operate over the state associated with the profile/key/peer-id. So only at the point when the profile is created can the service manager be informed that there needs to be a service which manages that daemon. And the parameterisation of the service depends on the outcome of the profile creation and the expectation is that after the creation there is a socket in a well-known location that reacts to IPC calls. Of course symmetrically when a profile is deleted the service manager should be informed of the need to remove the service, but that is not important to the point. Does that help illustrate the issue here? Otherwise I can try to draw something up to contextualise what I'm trying to get at.

Reply:
Taking the systemd version of this as an example there would need to be a unit somewhere describing the socket activated service for each profile. This unit would be created after the profile key and storage are created so that these parameters can be instantiated in the unit. One implication of this is that either you have to `sudo` to create a profile, or you use `systemd --user` etc.

Reply:
My example above assumes a user unit as it fits the model quite neatly.

Reply:
Yes yes user unit / Launch*Agent*. Note that it is still a good idea to revoke your own write permissions for security reasons, requiring sudo for modifications

Reply:
@FintanH Gentle ping to see if there is a common understanding of the interplay between profile management and daemon unit files. :dog:

Reply:
This is now capture as well in #722

Reply:
Just re-read this thread. Is the summary that unit files should be stored in a profile? And should be part of the `rad profile` setup logic? :)

Reply:
> Just re-read this thread. Is the summary that unit files should be stored in a profile? And should be part of the `rad profile` setup logic? :)

Where exactly they should be stored we can suss out in the RFC mentioned in #722, but generally speaking yes this is the takeaway.

Reply:
Happy for you to mark this as resolved from my side at least :)

Reply:
Is our common understanding that by facilities, we mean that our own tooling is able to perform the platform-specific (re-)configuration? As in, wrappers for systemd incantations.

Maybe we can say this more explicitly then :) It reads a little handwavy.

Reply:
abb9f67